### PR TITLE
Panorama - new commands for upload content usecase

### PR DESCRIPTION
**Pull request by abaumgarten**
## Status
- [] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
related: https://github.com/demisto/content/pull/10959
related: https://github.com/demisto/content/pull/9150
related: https://github.com/demisto/content/pull/10879

## Description
@apekarovsky's pack description: This content pack is set to be a "toolbox" for organisations with "Air Gap" networks. There are currently two use-cases in this "toolbox".
One is indicators "reputation" query from internal , air gapped network. This is done by getting the indicators across the "Air Gap" by using unidirectional diode that can pass UDP traffic. The answer will return as a JSON Feed file and will be transferred back in via another unidirectional diode that can pass files.
The second use-case is automating "Offline Panorama content updates" like AV, Apps and threats or even Wildfire content updates. External network Panorama will pull the content updates forward them to XSOAR. Then they will be checked for threats and moved by a unidirectional diode into the internal network where they will be automatically installed on the Panorama server and pushed to the firewalls.

## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 1]
```diff
@@ -4564,7 +4564,59 @@ script:
     - contextPath: Panorama.UserIDAgents.Disabled
       description: Whether the user-ID agent is disbaled.
       type: String
-  dockerimage: demisto/python3:3.9.2.17246
+  - name: panorama-upload-content-update-file
+    arguments:
+    - name: entryID
+      required: true
+      description: Entry id of the file to upload
```
Comment (bakatzir) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4571]:
```suggestion
      description: Entry ID of the file to upload.
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 2]
```diff
@@ -4564,7 +4564,59 @@ script:
...
+      type: string
+    - contextPath: Panorama.Content.Upload.Message
+      description: Content upload message.
+      type: string
+    description: Uploads content file to Panorama
+  - name: panorama-install-file-content-update
+    arguments:
+    - name: version_name
+      required: true
+      description: Update file name to be installed on Panorama
```
Comment (bakatzir) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4592]:
```suggestion
      description: Update file name to be installed on PAN-OS.
```

File: Tests/conf.json [hunk 3]
```diff
@@ -908,6 +908,13 @@
             "fromversion": "5.5.0",
             "timeout": 1600
         },
+        {
+            "integrations": "Panorama",
+            "instance_names": "palo_alto_panorama",
```
Comment (bakatzir) [inline Tests/conf.json:913]: if these commands are available on the FW as well, please also add an entry on the `palo_alto_firewall` or `palo_alto_firewall_9.0`

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 4]
```diff
@@ -4564,7 +4564,59 @@ script:
...
+      - anti-virus
+      - content
+      description: The file type
+    - name: skip_validity_check
+      required: true
+      auto: PREDEFINED
+      predefined:
+      - "yes"
+      - "no"
+      description: Skips file validity check with PANW update server. Use this option
```
Comment (bakatzir) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4607]:
```suggestion
      description: Skips file validity check with PAN-OS update server. Use this option
```

**Review by bakatzir — COMMENTED** [review]
- If these commands work on PAN-OS FW, please make sure the TPB runs on a FW instance as well. else - add to the desc that it is only available on Panorama.
- when running a content installation you get a `JobID` which is pending. why not poll on it until it is done in a playbook? how would else the customer know if the content update finished successfully or failed.

**Comment by apekarovsky** [comment]
@bakatzir 
1. The new commands were tested on Panorama and on FW (tested on v. 9)
2. That's already in the flow. The test playbook includes "generic polling" to check if the content update finished successfully or failed. That's true also for the content pack playbook and for the future use by the customer.

**Comment by bakatzir** [comment]
> @bakatzir
> 
> 1. The new commands were tested on Panorama and on FW (tested on v. 9)
> 2. That's already in the flow. The test playbook includes "generic polling" to check if the content update finished successfully or failed. That's true also for the content pack playbook and for the future use by the customer.

1. so add an entry that this TPB will run on the FW as well.
2. TPB is great. I'm talking about the need to add OOTB Playbook that will download the content and install it. then will poll on the jobID status until will change from `Pending` to `Success` - which I did not see in the code.

**Comment by apekarovsky** [comment]
Automatic download of the content file by XSOAR is currently not possible due to limitations on the PANW download portal. We need the customer to download the file and place it in a folder , we pick it up from there (also we are talking about "offline" content file upload). That's exactly what my "content pack" use case is about. You did not see the OOTB PB as @abaumgarten and I planning to submit it separately from the PANW integration enhancements.

**Comment by bakatzir** [comment]
> Automatic download of the content file by XSOAR is currently not possible due to limitations on the PANW download portal. We need the customer to download the file and place it in a folder , we pick it up from there (also we are talking about "offline" content file upload). That's exactly what my "content pack" use case is about. You did not see the OOTB PB as @abaumgarten and I planning to submit it separately from the PANW integration enhancements.

OK. that's a must be documented flow in the readme.
@abaumgarten Feel free to set a short demo with Anar and glicht on this.

**Comment by abaumgarten** [comment]
> > Automatic download of the content file by XSOAR is currently not possible due to limitations on the PANW download portal. We need the customer to download the file and place it in a folder , we pick it up from there (also we are talking about "offline" content file upload). That's exactly what my "content pack" use case is about. You did not see the OOTB PB as @abaumgarten and I planning to submit it separately from the PANW integration enhancements.
> 
> OK. that's a must be documented flow in the readme.
> @abaumgarten Feel free to set a short demo with Anar and glicht on this.

1. I have added an entry that this TPB will run on the FW as well.
2. @bakatzir do you suggest generating a readme file now? or after the docs-review?
3. I will schedule a demo, I think that if @apekarovsky will run this session it will be the best as he is the owner of this use case.

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 5]
```diff
@@ -6930,6 +6928,89 @@ def initialize_instance(args: Dict[str, str], params: Dict[str, str]):
...
+def panorama_upload_content_update_file_command(args: dict):
+    category = args.get('category')
+    entry_id = args.get('entryID')
+    file_path = demisto.getFilePath(entry_id)['path']
+    file_name = demisto.getFilePath(entry_id)['name']
+
+    try:
+        shutil.copy(file_path, file_name)
+    except Exception:
+        raise Exception('Failed to prepare file for upload.')
```
Comment (guykeller) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6940]: Why would a simple copying operation fail? And if it does fail, what insight will this error message yield?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 6]
```diff
@@ -6930,6 +6928,89 @@ def initialize_instance(args: Dict[str, str], params: Dict[str, str]):
...
+
+    with open(file_name, 'rb') as file:
+        files = {'file': file}
```
Comment (guykeller) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6943]: What is the purpose of this variable? unclear from the name IMO

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py [hunk 7]
```diff
@@ -6930,6 +6928,89 @@ def initialize_instance(args: Dict[str, str], params: Dict[str, str]):
...
+                                 outputs_prefix="Panorama.Content.Upload",
+                                 outputs_key_field="Status",
+                                 outputs=content_upload_info)
+
+    shutil.rmtree(file_name, ignore_errors=True)
+    return results
+
+
+@logger
+def panorama_install_file_content_update(version: str, category: str, validity: str):
```
Comment (guykeller) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.py:6962]: Adding a link to the API in a docstring could be helpful here, if possible

**Review by guykeller — CHANGES_REQUESTED** [review]
Pending demo

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 6 above)
Comment (abaumgarten) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: DONE

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 5 above)
Comment (abaumgarten) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py]: DONE

File: Packs/PAN-OS/Integrations/Panorama/Panorama.py (hunk 7 above)
Comment (abaumgarten) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.py:6958]: Done

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 8]
```diff
@@ -4564,7 +4564,59 @@ script:
...
+      required: true
+      description: Entry ID of the file to upload.
+    - name: category
+      required: true
+      auto: PREDEFINED
+      predefined:
+      - wildfire
+      - anti-virus
+      - content
+      description: The file type
```
Comment (amshamah419) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4579]: Is this the file type or the category of the file?

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 9]
```diff
@@ -4564,7 +4564,59 @@ script:
...
+    outputs:
+    - contextPath: Panorama.Content.Upload.Status
+      description: Content upload status.
+      type: string
+    - contextPath: Panorama.Content.Upload.Message
+      description: Content upload message.
+      type: string
+    description: Uploads content file to Panorama
```
Comment (amshamah419) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4587]:
```suggestion
    description: Uploads a content file to Panorama.
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 10]
```diff
@@ -4564,7 +4564,59 @@ script:
...
+      required: true
+      description: Update file name to be installed on PAN-OS.
+    - name: category
+      required: true
+      auto: PREDEFINED
+      predefined:
+      - wildfire
+      - anti-virus
+      - content
+      description: The file type
```
Comment (amshamah419) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4600]: Same here

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml [hunk 11]
```diff
@@ -4564,7 +4564,59 @@ script:
...
+    - name: skip_validity_check
+      required: true
+      auto: PREDEFINED
+      predefined:
+      - "yes"
+      - "no"
+      description: Skips file validity check with PAN-OS update server. Use this option
+        for networks disconnected from the Internet and only if you trust the content
+        file.
```
Comment (amshamah419) [inline Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4609]:
```suggestion
      description: Skips the file validity check with the PAN-OS update server. Use this option
        for air-gapped networks and only if you trust the content file.
```

File: Packs/PAN-OS/Integrations/Panorama/Panorama.yml (hunk 8 above)
Comment (abaumgarten) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml:4579]: @apekarovsky Please see that comment

Comment (abaumgarten) [reply Packs/PAN-OS/Integrations/Panorama/Panorama.yml]: Checked with @apekarovsky , should be a category of the content.
Changed.

**Comment by abaumgarten** [comment]
I removed the TPB from conf.json as we cannot upload the content file automatically (we talked about it during the demo session). The TPB is still available as part of the pack for future changes and testing.

**Review by guykeller — APPROVED** [review]
Looks fine to me, but pending @bakatzir approval as he attended the demo

**Review by bakatzir — APPROVED** [review]

**Pull request merged**